tree/forest/v7/inc/ROOT/RField.hxx

```diff
@@ -0,0 +1,760 @@
+/// \file ROOT/RField.hxx
+/// \ingroup Forest ROOT7
+/// \author Jakob Blomer <jblomer@cern.ch>
+/// \date 2018-10-09
+/// \warning This is part of the ROOT 7 prototype! It will change without notice. It might trigger earthquakes. Feedback
+/// is welcome!
+
+/*************************************************************************
+ * Copyright (C) 1995-2015, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT7_RField
+#define ROOT7_RField
+
+#include <ROOT/RColumn.hxx>
+#include <ROOT/RColumnElement.hxx>
+#include <ROOT/RField.hxx>
+#include <ROOT/RFieldValue.hxx>
+#include <ROOT/RForestUtil.hxx>
+#include <ROOT/RStringView.hxx>
+#include <ROOT/RVec.hxx>
+#include <ROOT/TypeTraits.hxx>
+
+#include <TGenericClassInfo.h>
+#include <TError.h>
+
+#include <algorithm>
+#include <iostream>
+#include <iterator>
+#include <memory>
+#include <string>
+#include <type_traits>
+#include <typeinfo>
+#include <vector>
+#include <utility>
+
+class TClass;
+
+namespace ROOT {
+namespace Experimental {
+
+class RCollectionForest;
+class RForestEntry;
+class RForestModel;
+class RFieldCollection;
+
+namespace Detail {
+
+class RPageStorage;
+
+// clang-format off
+/**
+\class ROOT::Experimental::RFieldBase
+\ingroup Forest
+\brief A field translates read and write calls from/to underlying columns to/from tree values
+
+A field is a serializable C++ type or a container for a collection of sub fields. The RFieldBase and its
+type-safe descendants provide the object to column mapper. They map C++ objects to primitive columns.  The
+mapping is trivial for simple types such as 'double'. Complex types resolve to multiple primitive columns.
+The field knows based on its type and the field name the type(s) and name(s) of the columns.
+*/
+// clang-format on
+class RFieldBase {
+   friend class ROOT::Experimental::RFieldCollection; // to change the field names when collections are attached
+private:
+   /// The field name is a unique within a tree and also the basis for the column name(s)
+   std::string fName;
+   /// The C++ type captured by this field
+   std::string fType;
+   /// The role of this field in the data model structure
+   EForestStructure fStructure;
+   /// A field on a trivial type that maps as-is to a single column
+   bool fIsSimple;
+
+protected:
+   /// Collections and classes own sub fields
+   std::vector<std::unique_ptr<RFieldBase>> fSubFields;
+   /// Sub fields point to their mother field
+   RFieldBase* fParent;
+   /// All fields have a main column. For collection fields, the main column is the index field. Points into fColumns.
```

Comment:
What is the principal column for a single (or couple) of 'free standing' float field(s)?